src/pp/if_parser.rs

```diff
@@ -0,0 +1,436 @@
+use crate::token::{Integer, PreprocessorError, Punct};
+
+use super::{Define, Location, MELexer, MacroProcessor, Step, StepExit, Token, TokenValue};
+use std::{collections::HashMap, rc::Rc, vec::IntoIter};
+
+struct IfLexer<'macros> {
+    tokens: IntoIter<Token>,
+    defines: &'macros HashMap<String, Rc<Define>>,
+}
+
+pub(super) struct IfParser<'macros> {
+    lexer: IfLexer<'macros>,
+    macro_processor: MacroProcessor,
+    location: Location,
+
+    parsing_if: bool,
+    carry: Option<Token>,
+}
+
+impl<'macros> IfParser<'macros> {
+    /// Builds a new IfParser that can be reused
+    ///
+    /// `parsing_if` indicates wether or not non defined macros should be
+    /// replaced with 0
+    pub fn new(
+        tokens: Vec<Token>,
+        defines: &'macros HashMap<String, Rc<Define>>,
+        location: Location,
+        parsing_if: bool,
+    ) -> Self {
+        IfParser {
+            lexer: IfLexer {
+                tokens: tokens.into_iter(),
+                defines,
+            },
+            macro_processor: MacroProcessor::default(),
+            location,
+
+            parsing_if,
+            carry: None,
+        }
+    }
+
+    fn raw_next(&mut self) -> Option<Token> {
```

Comment:
Could we have a comment that explains what the difference is between `raw_next`, `expect_next` and `next`? I think most of the complexity here comes from the handling of `defined`'s argument with raw tokens so we should call that out explicitly.